### ISSUE TITLE
Google Provider TODO: Handle OIDC response body error

### ISSUE THREAD

**Issue by INIRU** [nextauthjs/next-auth#10731]
### Provider type

Google

### Environment

```
 System:
    OS: macOS 14.4.1
    CPU: (11) arm64 Apple M3 Pro
    Memory: 127.47 MB / 18.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.7.3 - ~/.nvm/versions/node/v21.7.3/bin/node
    npm: 10.5.0 - ~/.nvm/versions/node/v21.7.3/bin/npm
  Browsers:
    Safari: 17.4.1
  npmPackages:
    @auth/mongodb-adapter: ^2.6.0 => 2.6.0 
    next: ^14.2.1 => 14.2.1 
    next-auth: ^5.0.0-beta.16 => 5.0.0-beta.16 
    react: ^18.2.0 => 18.2.0 
```

### Reproduction URL

https://github.com/TodayWantLook/TWL-NextJS/tree/developer/iniru

### Describe the issue

<img width="1165" alt="스크린샷 2024-04-26 오전 12 33 15" src="https://github.com/nextauthjs/next-auth/assets/75632393/f13e88a4-cdca-4e22-adf4-a7ceba61799b">
<img width="1165" alt="스크린샷 2024-04-26 오전 12 33 17" src="https://github.com/nextauthjs/next-auth/assets/75632393/838eb3bc-8742-47ce-93f9-299adf306363">
<img width="1165" alt="스크린샷 2024-04-26 오전 12 33 22" src="https://github.com/nextauthjs/next-auth/assets/75632393/4ed9d2fa-400f-4732-ae38-4946d5980bd0">

I seem to be having some issues getting the callbackurl from Google. I have attached the error in the picture above.

### How to reproduce

```dosini
//env.local

DB_NAME = DB_NAME
DB_URL = DB_URL

NEXTAUTH_URL = https://https://todaywantlook.vercel.app
AUTH_SECRET = LrKOg/MhErytOm35LY7grMi+/RdTCbUgbuezBOqdAhw=

AUTH_GOOGLE_ID = ''
AUTH_GOOGLE_SECRET = ''
```
<img width="550" alt="스크린샷 2024-04-26 오전 12 38 10" src="https://github.com/nextauthjs/next-auth/assets/75632393/c5e47d44-0cb3-498d-83a6-e96e6d5356c5">

After making the above settings, you should see a prompt to check the Server Error Log when you arrive at the login page.

### Expected behavior

The callbackurl should be routed correctly and you should be able to receive the login user information to the callbackurl normally.

**Comment by murilobd** [comment]
I'm having the same issue (I've just tried with beta 17 and still the same).

**Comment by ndom91** [comment]
So I'm not 100% sure about this, but google is setup as an OIDC provider which will auto-detect all the required URLs from their `.well-known` oIDC manifest, but you've hardcoded the authorihzation URL.

You can pass the same additional parameters that yuo're appending like this:

```ts
 providers: [
   GoogleProvider({
     authorization: {
       params: {
         prompt: "consent",
         access_type: "offline",
         response_type: "code"
       }
     }
   })
 ],
```

In addition, you can drop the `NEXTAUTH_URL` environment variable. It should be auto-detected on Vercel. If that doesn't help, the up-to-date version is called `AUTH_URL` btw

**Comment by INIRU** [comment]
Oh, thank you for helping me with my code. I'll check it out.

**Comment by murilobd** [comment]
I'm running the app on AWS Amplify (not sure on how different is from Vercel) and I noticed that, when I pass the AUTH_REDIRECT_PROXY_URL env variable, it succeeds on the login/register but it still redirects to localhost:3000.

**Comment by INIRU** [comment]
> So I'm not 100% sure about this, but google is setup as an OIDC provider which will auto-detect all the required URLs from their `.well-known` oIDC manifest, but you've hardcoded the authorihzation URL.
> 
> You can pass the same additional parameters that yuo're appending like this:
> 
> ```ts
>  providers: [
>    GoogleProvider({
>      authorization: {
>        params: {
>          prompt: "consent",
>          access_type: "offline",
>          response_type: "code"
>        }
>      }
>    })
>  ],
> ```
> 
> In addition, you can drop the `NEXTAUTH_URL` environment variable. It should be auto-detected on Vercel. If that doesn't help, the up-to-date version is called `AUTH_URL` btw


I tried the fix as you suggested, but it didn't fix the problem.
When I run it locally with `npm run dev` it logs in fine, but mysteriously when I push it up to the server, next-auth gives me an error.

**Comment by ndom91** [comment]
@INIRU can you share that next-auth error?

@murilobd the `AUTH_REDIRECT_PROXY_URL` option is designed for proxying preview deploy's through your prod deploy so you can login with Google (for example) on your preview deploys (or other apps), see: https://authjs.dev/getting-started/deployment#securing-a-preview-deployment

**Comment by ndom91** [comment]
@murilobd have yuo updated your callbackUrl in the Google / whatever OAUth provider dashboard? That's what tells the app where to redirect back to by default

**Comment by INIRU** [comment]
@ndom91 

```bash
[GET][xxx] /api/auth/callback/google
error { error: 'invalid_client', error_description: 'Unauthorized' }
```

```bash
[GET][xxx] /api/auth/callback/google
[auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
```

```bash
[GET][xxx] /api/auth/callback/google
[auth][cause] Error: TODO: Handle OIDC response body error
    at ib (/var/task/.next/server/chunks/168.js:393:28075)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async iR (/var/task/.next/server/chunks/168.js:393:33995)
    at async iW (/var/task/.next/server/chunks/168.js:393:45324)
    at async iq (/var/task/.next/server/chunks/168.js:393:50193)
    at async /var/task/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:34801
    at async eS.execute (/var/task/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:25910)
    at async eS.handle (/var/task/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:36055)
    at async ei (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:16:25466)
    at async es.responseCache.get.routeKind (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:1026)
```

```bash
[GET][xxx] /api/auth/callback/google
[auth][details][0m: {
  "provider": "google"
}
```

```bash
[GET][500] /api/auth/error
/api/auth/error?error=Configuration&nxtPnextauth=error status=500
```

**Comment by ndom91** [comment]
From that it seems like your `clientId` is either incorrect (copy and paste error maybe?) or just not being picked up. In order to infer it automatically, and not have to manually pass it to the `Google()` provider in your configuration, the environment variables have to be named like this:

```
AUTH_GOOGLE_ID
AUTH_GOOGLE_SECRET
```

Do you have that in place in Vercel?

**Comment by murilobd** [comment]
> @murilobd the AUTH_REDIRECT_PROXY_URL option is designed for proxying preview deploy's through your prod deploy so you can login with Google (for example) on your preview deploys (or other apps), see: https://authjs.dev/getting-started/deployment#securing-a-preview-deployment

Yes, I'm aware of that. But I thought maybe giving a try to bypass the "https://localhost:3000" and it worked partially. It stopped being redirected to localhost with queryParams error=Configuration (as it's happening with @INIRU )

> @murilobd have yuo updated your callbackUrl in the Google / whatever OAUth provider dashboard? That's what tells the app where to redirect back to by default

Yes. I've triple checked the callbackUrl to be the good one (I even created a new credential without the localhost just to be sure).

What I've just tried was to set the `redirect()` callback and it's working fine when the login succeeds. But, when it fails (in case when the account is not linked with the oauth provider), it redirects to https://localhost:3000/login?error=OAuthAccountNotLinked

```
redirect({ url, baseUrl }) {
			const finalUrl = new URL(url, process.env.APP_URL || baseUrl).href;
			console.log(
				" ------------ xxxxxx redirect xxxxxxx --------------- ",
				"url: " + url,
				"baseUrl: " + baseUrl,
				"finalUrl: " + finalUrl
			);
			return finalUrl;
		}
```

The `APP_URL` is an env variable I created with the good hostname.
From the console.log, indeed, the baseUrl is being considered as https://localhost:3000

![image](https://github.com/nextauthjs/next-auth/assets/4275011/cef3df93-96e4-4801-9e98-457fb9d678e4)

**Comment by murilobd** [comment]
@ndom91 just tried removing the AUTH_REDIRECT_PROXY_URL and it fails right after starting the auth process and redirects to: https://localhost:3000/api/auth/error?error=Configuration

**Comment by INIRU** [comment]
> From that it seems like your `clientId` is either incorrect (copy and paste error maybe?) or just not being picked up. In order to infer it automatically, and not have to manually pass it to the `Google()` provider in your configuration, the environment variables have to be named like this:
> 
> ```
> AUTH_GOOGLE_ID
> AUTH_GOOGLE_SECRET
> ```
> 
> Do you have that in place in your environment?

@ndom91 
Yes, It's already set up and passes well to Google OAuth. The error seems to be in the CallbackRouting part.

**Comment by ndom91** [comment]
@murilobd the correct environment variable is `AUTH_URL`, give that a shot. The `AUTH_REDIRECT_PROXY_URL` may have avoided the initial error, but it enables a whoel different flow inside the core, it's not what you want. Please remove it and try `AUTH_URL` instead of `APP_URL` :pray:

**Comment by ndom91** [comment]
> @ndom91 Yes, It's already set up and passes well to Google OAuth. The error seems to be in the CallbackRouting part.

Hmm well one way or another, google is telling us that the `client_id` it's receiving from you is invalid. I'd check the Google OAuth dashboard, maybe there's some more info there or in the worst case, I'd just generate a new clientId/clientSecret from Google

**Comment by murilobd** [comment]
> The error seems to be in the CallbackRouting part.

I do suspect that's it too!

**Comment by INIRU** [comment]
![image](https://github.com/nextauthjs/next-auth/assets/75632393/39c5e27a-d825-4a17-a7c7-9a4e3eb49835)

my vercel env settings

**Comment by murilobd** [comment]
> @murilobd the correct environment variable is `AUTH_URL`, give that a shot. The `AUTH_REDIRECT_PROXY_URL` may have avoided the initial error, but it enables a whoel different flow inside the core, it's not what you want. Please remove it and try `AUTH_URL` instead of `APP_URL` 🙏

When I set AUTH_URL, I have this error, simply by trying to access the webapp (and that's why I create the APP_URL)

![image](https://github.com/nextauthjs/next-auth/assets/4275011/db8604b7-2205-4e31-a9b9-17d157cdfe1c)

**Comment by murilobd** [comment]
@ndom91 do you know if it's normal, when running the app on aws amplify, to have the network set as localhost?
![image](https://github.com/nextauthjs/next-auth/assets/4275011/c050d65f-09d0-407a-be16-86c7c54b8171)

I'm asking this because, maybe all the redirects to `localhost` are because of that?

**Comment by ndom91** [comment]
Yeah so that's the default thing your node process running next.js will expose. So your node process is running on a VM or container or whatever in Amplify and its exposing the applicaiotn at `localhost:3000`. Then Amplify will put a reverse proxy in front of that node process which will terminal SSL and accept traffic at `https://app-123.aws.amplify.com` (or whatever).

That's why your main application doesn't know anythign about the final URL actually and you need to tell it via `AUTH_URL`. But also, that reminds me, you'll also need to enable the `trustHost` configuration option to tell next.js / auth.js to trust the URL that that reverse proxy tells the node process (via the `X-Forwarded-Host` header) ([Docs](https://authjs.dev/getting-started/deployment#auth_trust_host))

So long story short, try also adding `trustHost: true` to your auth.js config, or adding the `AUTH_TRUST_HOST=true` env var to your environment

**Comment by Nuno111** [comment]
Getting the same issue here.

I've got the same setup for github and google, github works fine but google gives me the invalid_client error and cause HANDLE OIDC response body error.

I've tripled checked, compared and searched by string, the clientId is 100% correct.

This is not new on my end, haven't been able to use google sign in  since i setup next auth

<img width="628" alt="image" src="https://github.com/nextauthjs/next-auth/assets/42497768/68fea584-446e-482b-80ae-9f8ef3629155">

**Comment by murilobd** [comment]
> Yeah so that's the default thing your node process running next.js will expose. So your node process is running on a VM or container or whatever in Amplify and its exposing the applicaiotn at localhost:3000. Then Amplify will put a reverse proxy in front of that node process which will terminal SSL and accept traffic at https://app-123.aws.amplify.com (or whatever) and pass that traffic onto the node process in the container at localhost:3000. SO yeah that part is fine 👍

Thank you for the very clear explanation on how it works! I suspected it was, indeed, a reverse proxy, but just would like to confirm that authjs wasn't doing anything related to that.

> So long story short, try also adding trustHost: true to your auth.js config, or adding the AUTH_TRUST_HOST=true env var to your environment

@ndom91 this is enabled. If I remove that configuration, I receive an error from authjs

**Comment by ndom91** [comment]
Hmm very weird, but that error object is coming from Google, so something seems wrong there as well. Maybe it's with what we send them, but the Google provider in our example app still works (https://next-auth-example.vercel.app/auth/signin) using `next-auth@5.0.0-beta.17`.

![image](https://github.com/nextauthjs/next-auth/assets/7415984/b037e297-4e00-47de-a982-1d10e1ec4c42)

![image](https://github.com/nextauthjs/next-auth/assets/7415984/a1afe009-00b4-437c-8891-cc4624745a6b)

@Nuno111 Can you provide some more some more details about your setup? `auth.ts` configuration, versions of pkg and next, etc. :pray:

**Comment by Nuno111** [comment]
Didn't fix it for me, same issue unfortunately!

**Comment by Nuno111** [comment]
@ndom91 

I dug a little deeper and it seems the codeGrantResponse is not authorized before making the proccesAuthorizationCodeOpenIDResponse function call,

<img width="407" alt="image" src="https://github.com/nextauthjs/next-auth/assets/42497768/c8813c2f-41ac-408f-9f04-0e400af6f83c">

<img width="775" alt="image" src="https://github.com/nextauthjs/next-auth/assets/42497768/18b0f1b3-dca7-4e02-becf-729031457673">

**Comment by ndom91** [comment]
@Nuno111 I see yuo're using `next-auth@5.0.0-beta.4`, have you tried the latest `-beta.17`?

**Comment by murilobd** [comment]
On `next-auth@5.0.0-beta.17` I'm still facing the same issue, being redirected to `https://localhost:3000` when authentication fails

**Comment by Nuno111** [comment]
@ndom91 @murilobd 

Bumping to .17 solved my issue.

I then ran into another problem which you can find here https://github.com/nextauthjs/next-auth/issues/9922

But basically i changed my NEXTAUTH_URL to NEXTAUTH_URL_INTERNAL and now i'm able to use google auth

Thanks for the support!

**Comment by murilobd** [comment]
I'm still having the same issue as before when user cancel the authentication with google. Instead of being redirected to the correct URL, it redirects to https://localhost:3000.

Running on next 2.4, next-auth@5.0.0-beta.17, prisma ^5.12.

Also, I'm using next-intl. Not sure if it plays a role in the bug I'm having, as I've mixed both in the middleware like that
```
import createIntlMiddleware from "next-intl/middleware";
import { auth as authMiddleware } from "./auth";
import { defaultLocale, locales } from "./i18n";

const intlMiddleware = createIntlMiddleware({
	locales,
	defaultLocale,
});

export default authMiddleware((req) => intlMiddleware(req));

export const config = {
	matcher: [
		"/((?!api|_next|favicon.ico|manifest.json|.*\\..*).*)",
	],
};
```

**Comment by codemaniac** [comment]
I'm still facing the issue using `Auth0` as my provider and using `@auth/sveltekit`!
I'm using this sample project: https://github.com/nextauthjs/sveltekit-auth-example
Tried reading through this thread to troubleshoot, but couldn't find anything useful! Please help!

### Environment
```
 System:
    OS: Windows 11
    CPU: Core i7
  Binaries:
    Node: v20.12.2
```

### package.json

```
"dependencies": {
    "@auth/sveltekit": "latest"
},
```

### .env.local

```
AUTH_AUTH0_ID=123abc
AUTH_AUTH0_SECRET=xyz
AUTH_AUTH0_ISSUER=https://dev-abcdefg.us.auth0.com/
AUTH_SECRET=123456
```

### src/auth.ts

```
import { SvelteKitAuth } from "@auth/sveltekit"
import Auth0 from "@auth/sveltekit/providers/auth0"


export const { handle, signIn, signOut } = SvelteKitAuth({
  trustHost: true,
  providers: [Auth0],
})
```

### src/hooks.server.ts

```
export { handle } from "./auth"
```

### Error Message

```
error { error: 'access_denied', error_description: 'Unauthorized' }
[auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle OIDC response body error
    at handleOAuth (file:///C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@auth/core/lib/actions/callback/oauth/callback.js:86:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.callback (file:///C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@auth/core/lib/actions/callback/index.js:22:41)
    at async AuthInternal (file:///C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@auth/core/lib/index.js:27:24)
    at async Module.Auth (file:///C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@auth/core/index.js:109:34)
    at async Module.respond (C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@sveltejs/kit/src/runtime/server/respond.js:330:20)
    at async file:///C:/Users/ashis/Documents/Stack/code/clients/sveltekit-auth-example-main/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:524:22
[auth][details]: {
  "provider": "auth0"
}
```

### Auth0 Settings

`Application Login URI = <blank>`
`Allowed Callback URLs = http://localhost:5173/auth/callback/auth0`
`Allowed Logout URLs = http://localhost:5173`
`Allowed Web Origins = http://localhost:5173`
`Allowed Origins (CORS) = http://localhost:5173`

**Comment by min-stelland** [comment]
> Yeah so that's the default thing your node process running next.js will expose. So your node process is running on a VM or container or whatever in Amplify and its exposing the applicaiotn at `localhost:3000`. Then Amplify will put a reverse proxy in front of that node process which will terminate and negotiate the SSL/TLS connection and accept traffic at `https://app-123.aws.amplify.com` (or whatever) and pass that traffic onto the node process in the container at `localhost:3000`. So yeah that part is fine 👍
> 
> But also, the fact that that reverse proxy is in the way, is why your application doesn't know anything about the final URL and you need to tell it via `AUTH_URL`. But also, that reminds me (!), you'll also need to enable the `trustHost` configuration option to tell next.js / auth.js to trust the URL that that reverse proxy tells the node process (via the `X-Forwarded-Host` header it appends to requests when handing the request off to the node process) ([Docs](https://authjs.dev/getting-started/deployment#auth_trust_host))
> 
> So long story short, try also adding `trustHost: true` to your auth.js config, or adding the `AUTH_TRUST_HOST=true` env var to your environment
> 
> **Sidenote**: In real life there's multiple reverse proxies / load balancers / CDN's in between you and the node process running your node.js app in an AWS container in amplify or whatever.

This is the solution that worked for me. Thank you very much!

**Comment by tatleung** [comment]
My setup is as follows:

- Nginx webserver exposed to the public with HTTPS enabled.
- NextJS 15 application running on a different machine than the Nginx web server.
- Use Google for OAuth2 authentication
- next-auth 5.0.0-beta.19

I initially got the same error reported in this ticket.  After I added AUTH_URL to .env.local with the value set to http://<web_server_host> and update the "Authorized redirect URIs" in Google to http://<web_server_host>/api/auth/callback/google, the problem does not occur anymore.  Notice even though my web server is using HTTPS, I have to use HTTP for the redirect url.  Using HTTPS did not work for me.

I have another application that is using next-auth 4.  The "Authorized redirect URIs" I set there uses HTTPS.  The webserver and NextJS app server setup for both applications are essentially identical.  There seems to be some difference between Next-Auth 4 and 5.

BTW, when I configure the app using http://localhost:3000 (not using webserver reverse proxy), it is also working correctly.  Hope this helps someone.

**Comment by mahbd** [comment]
I don't know how and why but in my case it got solved after changing AUTH_SECRET. 

My project was running just fine in local machine but was not working in vercel. Then I copied my local machine AUTH_SECRET and now it is working. Maybe AUTH_SECRET was badly formatted or something else I don't know.

**Comment by mariomeissner** [comment]
Leaving a note here for others just in case it's helpful.

I had a functioning auth setup with next-auth beta version 19 deployed on Vercel.
When I moved to Railway, the issue described in this ticket started happening. The likely explanation is the way that Railway exposes the endpoint via proxy, such that it thinks it should use `localhost` when in fact it should be using the public domain URL.

Setting `NEXT_AUTH=https://my-website.com/api/auth/` in Railway fixed the issue for me!

**Comment by zhm** [comment]
I just spent way too long troubleshooting this `Error: TODO: Handle OIDC response body error` error and the problem ended up being I missed a double quote in the env file for the `AUTH_GOOGLE_SECRET` value. Face palm!
